src/realm/parser/query_builder.cpp

```diff
@@ -128,21 +128,20 @@ struct PropertyExpression
     size_t col_ndx;
     DataType col_type;
 
-    PropertyExpression(Query &query, const std::string &key_path_string)
+    PropertyExpression(Query &query, const std::string &key_path_string, const std::string& object_type)
     {
         KeyPath key_path = key_path_from_string(key_path_string);
         TableRef cur_table = query.get_table();;
         for (size_t index = 0; index < key_path.size(); index++) {
             size_t cur_col_ndx = cur_table->get_column_index(key_path[index]);
-            StringData table_name = cur_table->get_name();
 
             precondition(cur_col_ndx != realm::not_found,
-                         util::format("No property '%1' on object of type '%2'", key_path[index], table_name));
+                         util::format("No property '%1' on object of type '%2'", key_path[index], object_type));
```

Comment:
Isn't this now using the incorrect type name when links are involved? If we've followed a link, we're no longer looking at the original type but the linked type, and it is that linked type whose name should appear in this error.